mol-loader/Pubchem.py

```diff
@@ -0,0 +1,69 @@
+"""
+Encapsulate reading and formatting of PubChem
+"""
+
+from Sdf import Sdf
+import json
+import os, os.path
+import rdkit.Chem
+
+class Pubchem(Sdf):
+    def __init__(self, path):
+        Sdf.__init__(self, path)
+
+    SYNONYM_PROPS = [
+        'PUBCHEM_IUPAC_OPENEYE_NAME',
+        'PUBCHEM_IUPAC_CAS_NAME',
+        'PUBCHEM_IUPAC_NAME',
+        'PUBCHEM_IUPAC_SYSTEMATIC_NAME',
+        'PUBCHEM_IUPAC_TRADITIONAL_NAME'
+    ]
+    SMILES_PROPS = [
+        'PUBCHEM_OPENEYE_CAN_SMILES',
+        'PUBCHEM_OPENEYE_ISO_SMILES'
+    ]
+
+    def mol_to_dict(self, mol):
+        """
+        Capture all the information from a PubChem molecule into a dictionary
+
+        mol_to_dict( (Pubchem)self, (rdkit.Chem.Mol)mol) -> dict
+        """
+        props = set(mol.GetPropNames())
+        # All the properties defined in the molecule, as-is
+        d = {p.lower() : self.cast(mol.GetProp(p)) for p in props}
+        d['id'] = 'https://pubchem.ncbi.nlm.nih.gov/compound/' + mol.GetProp('PUBCHEM_COMPOUND_CID')
+
+        smiles = [mol.GetProp(p) for p in self.SMILES_PROPS if p in props]
+        smiles.append(rdkit.Chem.MolToSmiles(mol)) # Original SMILES
+        # Attempt normalization; this may fail if there are ... oddities ...
+        # in the molecule.
+        rdkit_mol = rdkit.Chem.MolFromSmiles(rdkit.Chem.MolToSmiles(mol)) # Canonicalizes
```

Comment:
Yeah, it will leave Hs in. You can replace the call to Chem.SanitizeMol()
with:
`rdkit_mol = Chem.RemoveHs(rdkit_mol)`
that will pull the Hs off and then sanitize things.


On Thu, Mar 3, 2016 at 4:49 PM, Nik Bates-Haus <notifications@github.com>
wrote:

> In mol-loader/Pubchem.py
> <https://github.com/nbateshaus/chem-search/pull/11#discussion_r54898622>:
>
> > +    def mol_to_dict(self, mol):
> > +        """
> > +        Capture all the information from a PubChem molecule into a dictionary
> > +
> > +        mol_to_dict( (Pubchem)self, (rdkit.Chem.Mol)mol) -> dict
> > +        """
> > +        props = set(mol.GetPropNames())
> > +        # All the properties defined in the molecule, as-is
> > +        d = {p.lower() : self.cast(mol.GetProp(p)) for p in props}
> > +        d['id'] = 'https://pubchem.ncbi.nlm.nih.gov/compound/' + mol.GetProp('PUBCHEM_COMPOUND_CID')
> > +
> > +        smiles = [mol.GetProp(p) for p in self.SMILES_PROPS if p in props]
> > +        smiles.append(rdkit.Chem.MolToSmiles(mol)) # Original SMILES
> > +        # Attempt normalization; this may fail if there are ... oddities ...
> > +        # in the molecule.
> > +        rdkit_mol = rdkit.Chem.MolFromSmiles(rdkit.Chem.MolToSmiles(mol)) # Canonicalizes
>
> I wasn't having good luck with SanitizeMol(mol) - e.g. it seemed to be
> leaving Hs in. I'll give it another whirl; I had a bunch of stuff
> in-flight, so it was probably me.
> Is SanitizeMol() that much faster than Mol->SMILES->Mol?
> This pipeline seems to be able to load Solr at a consistent rate of about
> 1K Mols/minute, which has Python and Java each consuming 1 cpu. I know
> Solr/Java can use more CPUs, so this seems Python-CPU-bound right now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nbateshaus/chem-search/pull/11/files#r54898622>.
>